src/unix/fs.c

```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
   return n;
 }
 
+static int uv__fs_opendir(uv_fs_t* req) {
+  DIR *dir_stream;
+  uv__dirent_t *dirent;
+  uv_dir_t *dir;
+  size_t len;
 
-static ssize_t uv__fs_readlink(uv_fs_t* req) {
-  ssize_t len;
-  char* buf;
+  assert(req);
 
-  len = pathconf(req->path, _PC_PATH_MAX);
+  dir_stream = opendir(req->path);
+  if (dir_stream == NULL)
+    return -1;
 
-  if (len == -1) {
-#if defined(PATH_MAX)
-    len = PATH_MAX;
-#else
-    len = 4096;
-#endif
+  len = uv__fs_direntry_size(dir_stream);
+  if (len == (size_t)-1)
+    return -1;
+
+  /*
+  * Allocate the space for each directory entry just once instead of
+  * once per directory entry.
+  */
+  dirent = malloc(len + 1);
+  if (dirent == NULL) {
+   errno = ENOMEM;
+   goto failed;
+  }
+
+  dir = malloc(sizeof(*dir));
+  if (dir == NULL) {
+    errno = ENOMEM;
+    goto failed;
+  }
+
+  dir->dir = dir_stream;
+  dir->dirent = dirent;
+
+  req->dir = dir;
+  req->ptr = (void*)req->dir;
```

Comment:
Without, you get this:

```
src/unix/fs.c:488:12: warning: assigning to 'void *' from 'const uv_dir_t *' (aka 'const struct uv_dir_s *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
  req->ptr = req->dir;
           ^ ~~~~~~~~
```

I'll see if there is a better approach.  _(I inherited some of this so bear with me)_